funsor/joint.py

```diff
@@ -155,31 +154,28 @@ def moment_matching_reduce(self, op, reduced_vars):
 
             # Moment-matching approximation.
             assert approx_vars and not exact_vars
-            discrete = self.discrete
-
+            gaussian = self.gaussian
+            discrete = self.discrete + gaussian.log_normalizer
             new_discrete = discrete.reduce(ops.logaddexp, approx_vars.intersection(discrete.inputs))
-            num_elements = reduce(ops.mul, [
-                self.inputs[k].num_elements for k in approx_vars.difference(discrete.inputs)], 1)
-            if num_elements != 1:
-                new_discrete -= math.log(num_elements)
 
-            gaussian = self.gaussian
             int_inputs = OrderedDict((k, d) for k, d in gaussian.inputs.items() if d.dtype != 'real')
             probs = (discrete - new_discrete).exp()
             old_loc = Tensor(cholesky_solve(gaussian.info_vec.unsqueeze(-1),
                                             gaussian._precision_chol).squeeze(-1),
                              int_inputs)
             new_loc = (probs * old_loc).reduce(ops.add, approx_vars)
-            old_cov = Tensor(sym_inverse(gaussian.precision), int_inputs)
+            old_cov = Tensor(cholesky_inverse(gaussian._precision_chol), int_inputs)
```

Comment:
@fritzo I think that you can compute `old_cov` first, then compute `old_loc` based on `old_cov`. For `new_precision`, I can't find a better identity to compute it. For `new_cov`, can we add `old_cov` and `outers` before applying `reduce(ops.add...)`? I don't know enough funsor to justify it.

Reply:
> you can compute old_cov first, then compute old_loc based on old_cov

`cholesky_solve` is more numerically stable than `cholesky_inverse(...).matmul()` and has same asymptotic complexity.

> can we add `old_cov` and outers before applying `reduce(ops.add...)`

It is cheaper to reduce first and add smaller things (even though it adds a tensor op).